code/datums/minimap.dm

```diff
@@ -0,0 +1,415 @@
+		"z_level" = z_level,
+		"save_location" = save_location,
+		"tile_offset_x" = tile_offset_x,
+		"tile_offset_y" = tile_offset_y,
+		"tile_width" = tile_width,
+		"tile_height" = tile_height,
+		"section_width" = section_width,
+		"section_height" = section_height,
+		"section_columns" = section_columns,
+		"section_rows" = section_rows,
+		"asset_paths" = json_encode(minimap_asset.assets),
+	)), "[save_location]/persistence.json")
+
+/datum/minimap_data/proc/try_load_persistence_data()
+	var/persistence_raw = file2text("[save_location]/persistence.json")
+	if(!length(persistence_raw))
+		return FALSE
+
+	var/list/persistence_data = json_decode(persistence_raw)
+	if(!length(persistence_data))
+		return FALSE
+	if(persistence_data["map_config_key"] != SSmapping.config.config_filename)
+		return FALSE
+	if(persistence_data["save_location"] != save_location)
+		return FALSE
+	if(persistence_data["git_commit"] != GLOB.revdata.commit)
+		return FALSE
+
+	z_level = persistence_data["z_level"]
+	save_location = persistence_data["save_location"]
+	tile_offset_x = persistence_data["tile_offset_x"]
+	tile_offset_y = persistence_data["tile_offset_y"]
+	tile_width = persistence_data["tile_width"]
+	tile_height = persistence_data["tile_height"]
+	section_width = persistence_data["section_width"]
+	section_height = persistence_data["section_height"]
+	section_columns = persistence_data["section_columns"]
+	section_rows = persistence_data["section_rows"]
+	minimap_asset = new /datum/asset/simple
+	minimap_asset.assets = json_decode(persistence_data["asset_paths"])
+	minimap_asset.register()
+	return TRUE
+
+/// Holds information used for generating a minimap.
+/datum/minimap_generation_cache
+	var/started_at //! When the generation of the minimap was queued
+	var/list/datum/minimap_generation_section/sections
+	var/datum/minimap_generation_section/current_section
+
+	var/datum/minimap_data/data //! The data for the minimap
+
+/datum/minimap_generation_section
+	var/section_x
+	var/section_y
+	var/icon/section_blank
+	var/list/section_turfs = list()
+
+/datum/controller/subsystem/minimap/Initialize()
+	for(var/z_level in SSmapping.levels_by_trait(ZTRAIT_STATION))
+		generate_minimap_for_z(z_level, allow_persistence_recovery = TRUE)
+
+	return SS_INIT_SUCCESS
+
+/datum/controller/subsystem/minimap/Recover()
+	minimaps_by_z_level = SSminimap.minimaps_by_z_level
+	to_generate = SSminimap.to_generate
+	generating = SSminimap.generating
+	current_section = SSminimap.current_section
+	current_run = SSminimap.current_run
+
+/datum/controller/subsystem/minimap/stat_entry(msg)
+	. = ..()
+	if(isnull(generating))
+		return .
+	return . + "S:[current_section.section_x],[current_section.section_y]|CR:[length(current_run)]"
+
+/// Generates a minimap for the given z level. Does not block and will generate the minimap in the background.
+/datum/controller/subsystem/minimap/proc/generate_minimap_for_z(z, allow_persistence_recovery = FALSE)
+	set waitfor = FALSE
```

Comment:
I greatly disagree with this, not only are you causing init work to be hidden from stats by causing this to run after init is "complete", this also makes all further subsystem initialization times inaccurate as you're doing work during their init slot.

Post init, pre game, period of the start of a round is for two purposes in my eyes:
- Avoid server costs as much as possible so that it can focus on clients changing preferences/characters/etc
- Incidental work caused by objects spawning in the world (This should be minimized but it's going to happen)

This subsystem should be doing work during it's given period and measured like any other. Yes it will slow down init, but if we just moved everything we could out of init in the same way as here then nothing would be gained except a second init phase that doesn't measure anything, and no init speedup either.

Reply:
no work is being done in this proc except calculations on minimap size and offset. All actual work is done inside /fire

Reply:
I needed to put the review comment somewhere, this isn't really about this particular line of code though I put it on something that probably won't be there after changes that would need to be done